File: backend/src/main/java/com/jllado/weightcontrol/service/UserProfileService.java
package com.jllado.weightcontrol.service;

import com.jllado.weightcontrol.api.dto.UserProfileDtos.UserProfileRequest;
import com.jllado.weightcontrol.domain.User;
import com.jllado.weightcontrol.repository.UserRepository;
import com.jllado.weightcontrol.util.DateTimes;
import jakarta.transaction.Transactional;
import java.time.LocalDate;
import org.springframework.stereotype.Service;

@Service
@Transactional
public class UserProfileService {

    private final UserRepository userRepository;

    public UserProfileService(UserRepository userRepository) {
        this.userRepository = userRepository;
    }

    public User update(User user, UserProfileRequest request) {
        validate(request);
        user.setBirthDate(request.birthDate());
        user.setHeightCm(request.heightCm());
        user.setSex(request.sex());
        user.setFitnessLevel(request.fitnessLevel());
        user.setTakesMedication(request.takesMedication());
        return userRepository.save(user);
    }

    private void validate(UserProfileRequest request) {
        if (request.birthDate() != null && request.birthDate().isAfter(LocalDate.now(DateTimes.USER_ZONE))) {
            throw new BadRequestException("Birth date cannot be in the future");
        }
    }
}

File: src/services/UserProfileService.js
import dayjs from 'dayjs';
import {get, put} from './api';
import UserProfile from '../model/UserProfile';

function toProfile(data) {
    return new UserProfile(data);
}

function toPayload(profile) {
    return {
        birthDate: profile.birthDate ? dayjs(profile.birthDate).format('YYYY-MM-DD') : null,
        heightCm: profile.heightCm,
        sex: profile.sex,
        fitnessLevel: profile.fitnessLevel,
        takesMedication: profile.takesMedication,
        typicalCaloriesPerDay: {...profile.typicalCaloriesPerDay}
    };
}

export default {
    async get() {
        return toProfile(await get('/profile'));
    },
    async save(profile) {
        return toProfile(await put('/profile', toPayload(profile)));
    }
}
